backend/assignments/2/src/main/java/com/kdu/caching/service/GeocodingService.java

```diff
@@ -0,0 +1,108 @@
+package com.kdu.caching.service;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.kdu.caching.dto.forwardgeocoding.ForwardGeocodingRequest;
+import com.kdu.caching.exceptions.GeocodingException;
+import com.kdu.caching.model.GeocodingData;
+import com.kdu.caching.model.ReverseGeocodingData;
+import com.kdu.caching.utils.GeocodingUtils;
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Value;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+public class GeocodingService {
+
+    public GeocodingService() {
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GeocodingService.class);
+
+    private RestTemplate restTemplate = new RestTemplate();
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    @Value("${api-key}")
+    private String apiKey;
+    @Value("${geocoding-url}")
+    private String geocodingUrl;
+    @Value("${reverse-geocoding-url}")
+    private String reverseGeocodingUrl;
+
+
+    @Cacheable(value = GeocodingUtils.FORWARD_GEOCODING_NAME, key = GeocodingUtils.FORWARD_GEOCODING_KEY_PREFIX)
+    public GeocodingData forwardGeocoding(String request) {
+        try {
+            // Build the URI using UriComponentsBuilder
+            String apiUrl = geocodingUrl + apiKey + "&query=" + request;
+
+            LOGGER.info("Constructed API URL: {}", apiUrl);
+
+            JsonNode apiResponse = restTemplate.getForObject(apiUrl, JsonNode.class);
+            LOGGER.info("API Response: {}", apiResponse);
+            try {
```

Comment:
Rather than adding multiple try-catch blocks. Wrap the logic in a method. Handle it there.
Single try-catch should be sufficient per function. If you need more, means you need to modularise more.